lighthouse-core/gather/computed/first-interactive.js

```diff
@@ -0,0 +1,206 @@
+ *      definition.
+ *    > The page responds to user input in a reasonable time on average, but it’s ok if this
+ *      response is not always immediate.
+ *
+ * First Interactive is defined as the first period after FMP of N-seconds that has no bad task
+ * clusters.
+ *
+ *    > t = time in seconds since FMP
+ *    > N = f(t) = 4 * e^(-0.045 * t) + 1
+ *      5 = f(0) = 4 + 1
+ *      3 ~= f(15) ~= 2 + 1
+ *      1 ~= f(∞) ~= 0 + 1
+ *    > a "bad task cluster" is a cluster of 1 or more long tasks with no more than 1s of idle time
+ *      between each task that does one of the following
+ *        > Starts within the first 5s after FMP.
+ *        > Spans more than 250ms from the start of the earliest task in the cluster to the end of the
+ *          latest task in the cluster.
+ *
+ * If this timestamp is earlier than DOMContentLoaded, use DOMContentLoaded as firstInteractive.
+ */
+class FirstInteractive extends ComputedArtifact {
+  get name() {
+    return 'FirstInteractive';
+  }
+
+  /**
+   * @param {number} t The time passed since FMP in miliseconds.
+   * @return {number}
+   */
+  static getRequiredWindowSizeInMs(t) {
+    const tInSeconds = t / 1000;
+    const exponentiationComponent = Math.exp(-.045 * tInSeconds);
+    return (4 * exponentiationComponent + 1) * 1000;
+  }
+
+  /**
+   * @param {!Array<{start: number, end: number}>} tasks
+   * @param {number} startIndex
+   * @param {number} windowEnd
+   * @return {!Array<{start: number, end: number, duration: number}>}
+   */
+  static getTaskClustersInWindow(tasks, startIndex, windowEnd) {
+    const clusters = [];
+
+    let previousTaskEndTime = -Infinity;
+    let currentCluster = null;
+
+    // Examine all tasks that could possibly be part of a cluster starting before windowEnd.
+    // Consider the case where window end is 15s, there's a 100ms task from 14.9-15s and a 500ms
+    // task from 15.5-16s, we need that later task to be clustered with the first so we can properly
+    // identify that main thread isn't quiet.
+    const clusteringWindowEnd = windowEnd + MIN_TASK_CLUSTER_PADDING;
+    const couldTaskBelongToCluster = task => task && task.start < clusteringWindowEnd;
+    for (let i = startIndex; couldTaskBelongToCluster(tasks[i]); i++) {
+      const task = tasks[i];
+
+      // if enough time has elapsed, we'll create a new cluster
+      if (task.start - previousTaskEndTime > MIN_TASK_CLUSTER_PADDING) {
+        if (currentCluster) {
+          clusters.push(currentCluster);
+        }
+
+        currentCluster = [];
+      }
+
+      currentCluster.push(task);
+      previousTaskEndTime = task.end;
+    }
+
+    if (currentCluster) {
+      clusters.push(currentCluster);
+    }
+
+    return clusters
+      // add some useful information about the cluster
+      .map(tasks => {
+        const start = tasks[0].start;
+        const end = tasks[tasks.length - 1].end;
+        const duration = end - start;
+        return {start, end, duration, tasks};
```

Comment:
any reason to keep `tasks` around? (not on the return type and looks like not used, so OK to drop?)